### PR TITLE
Boost.python requires numpy. Fix function _run_python_script().

### PR DESCRIPTION
**Pull request by wdobbe**
Specify library name and version:  **boost/all_versions**

**First fix:**
Function `_run_python_script` was broken when provided script returns empty output.
And on Windows the script that  `_python_abiflags(self)` sends to `_run_python_script` does return no output. As a result the `_run_python_script` returned something like
```
 ----Running------
> "python" -c "from __future__ import print_function; import sys; print(getattr(sys, 'abiflags', ''))"
-----------------
```
instead of an empty string. Result was the following error message when building the boost recipe with option `without_python=False` even though python libraries were present: 

`ERROR: boost/1.76.0: Invalid configuration: couldn't locate python libraries - make sure you have installed python development files`

**Second fix:**
Build of Boost.python library will fail if python package numpy is not installed. Check for presence of numpy in validate().

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Nekto89** [comment]
Is numpy dependency new for Boost.Python? I'm currently using 1.74.0 that was built without conan and without numpy.

File: recipes/boost/all/conanfile.py [hunk 1]
```diff
@@ -385,6 +386,13 @@ def validate(self):
                     if min_compiler_version is not None:
                         if tools.Version(self.settings.compiler.version) < min_compiler_version:
                             raise ConanInvalidConfiguration("Boost.Math requires a C++11 capable compiler")
+                            
+        if not self.options.without_python:
+            #Boost python library build will fail if python package numpy is not installed
+            try:
+                import numpy
```
Comment (Nekto89) [inline recipes/boost/all/conanfile.py:393]: what if we want to use different Python that was given through "python_executable" for example?

**Comment by wdobbe** [comment]
> Is numpy dependency new for Boost.Python? I'm currently using 1.74.0 that was built without conan and without numpy.

From [this vcpkg issue](https://github.com/microsoft/vcpkg/issues/5732) it seems that Boost should check during build if numpy is installed or not and decides whether to build boost numpy library based on that result. If it builds boost numpy this will result in an extra library (libboost_numpy39.lib on Windows).
So the question is why Boost starts to build the numpy library on my systems even when numpy package is not installed.

In general: if the check for python package numpy in this PR is a problem, then better to remove that part and think about how we can fix it properly later.

The other patch is more important, because it completely messes up the boost python build on Windows and it takes quite a bit of time and debugging to find out what is going wrong.

**Comment by wdobbe** [comment]
ps: I don't understand why the license/cla is shown as 'not signed yet'. I have already signed it multiple times and have submitted PRs before. I signed the CLA again yesterday, but no result yet.

**Comment by prince-chrismc** [comment]
The commits is @winfriedd ... you need to sign the CLA with that account

**Comment by wdobbe** [comment]
Thanks, I didn't even know I had that other Github account. I will close that asap.

File: recipes/boost/all/conanfile.py (hunk 1 above)
Comment (prince-chrismc) [reply recipes/boost/all/conanfile.py:393]: You should use the run python script helper to test for this

Comment (SSE4) [reply recipes/boost/all/conanfile.py:393]: the problem is with `CONAN_PRINT_RUN_COMMANDS`(well-known issue: https://github.com/conan-io/conan/issues/8319 https://github.com/conan-io/conan/issues/6752)
seems like it's not going to be fixed, so I guess we have to add a workaround to `_run_python_script`

Comment (SSE4) [reply recipes/boost/all/conanfile.py:393]: actually, there is already a workaround for that: https://github.com/conan-io/conan-center-index/pull/5643/files#diff-cc14160728d1ad321c1444ffad82615139181e681f4acb7dfd045a005cc8bd66R487
seems like it doesn't always work?

File: recipes/boost/all/conanfile.py [hunk 2]
```diff
@@ -478,6 +486,8 @@ def _run_python_script(self, script):
         # Conan is broken when run_to_output = True
         if "\n-----------------\n" in output:
             output = output.split("\n-----------------\n", 1)[1]
+        elif re.search('\n----+$', output) != None:
```
Comment (SSE4) [inline recipes/boost/all/conanfile.py:489]: I am curious, why above two lines didn't strip bad lines for you?
I tried locally, and for me it works correctly on Windows
but seems like in your case, it's still returning an output with dashes not removed...
maybe you can dump an unmodified output before line 486?

Comment (SSE4) [reply recipes/boost/all/conanfile.py:489]: /cc @madebr as seems like he did an original workaround

Comment (wdobbe) [reply recipes/boost/all/conanfile.py:489]: @SSE4 The output of the `self.run()` function ends with `\n-----------------\n<script output>` if the script does return output. But when the script returns nothing the output of the `self.run()` function ends with `\n-----------------` so the output is not split.
There are of course alternative implementations, we could also split at `\n-----------------` and then strip a possible leading `\n` from the result.

On my Windows machine the script that _python_abiflags(self) runs gives no (empty) output, that's were the build messed up.

Comment (SSE4) [reply recipes/boost/all/conanfile.py:489]: oh, I see now, and can reproduce: the `strip()` call removes `\n` at the end, so output doesn't match `\n-----------------\n` any longer.
the following fixes abi flags for me:
```
index 359e4a3..98caa84 100644
--- a/recipes/boost/all/conanfile.py
+++ b/recipes/boost/all/conanfile.py
@@ -501,10 +501,12 @@ class BoostConan(ConanFile):
         except ConanException:
             self.output.info("(failed)")
             return None
-        output = output.getvalue().strip()
+        output = output.getvalue()
         # Conan is broken when run_to_output = True
         if "\n-----------------\n" in output:
             output = output.split("\n-----------------\n", 1)[1]
+        output = output.strip()
```

Comment (madebr) [reply recipes/boost/all/conanfile.py:489]: @SSE4 's implementation looks easier to understand.
Let's hope switching to conan 2.0 will drop the need for this workaround.

Comment (wdobbe) [reply recipes/boost/all/conanfile.py]: Confirm, that works also for me. I changed the code accordingly.

File: recipes/boost/all/conanfile.py (hunk 1 above)
Comment (wdobbe) [reply recipes/boost/all/conanfile.py]: Even when python numpy is installed, when I build the boost package with conan create using VS2019 I get unresolved symbol link errors on boost.numpy for the numpy.cpp executable in the test package.
So better to fix the numpy problems in a separate PR, I removed the numpy check that I added from this PR.

**Review by SSE4 — APPROVED** [review]

**Comment by SSE4** [comment]
the following code could be used to detect numpy (better in another PR):
```
    @property
    def _has_numpy(self):
        """
        obtain the result of the "sysconfig.get_python_inc()" call
        :return: result of the "sysconfig.get_python_inc()" execution
        """
        return self._run_python_script("from __future__ import print_function; "
                                       "import numpy; "
                                       "print(numpy.__version__)")
```

**Comment by wdobbe** [comment]
Thanks. I noticed that when building the test_package with VS2019 linking the numpy.cpp executable fails with unresolved synbol errors to boost numpy. So I removed numpy stuff from this PR, needs more investigation.

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**